### PR TITLE
testing(dot): migrate dot/sync tests to integration tests

### PR DESCRIPTION
**Pull request by edwardmack**
## Changes

- Migrate existing non-mocked tests to integration tests, prepend file with build tag.
- Change filenames to `xx_integration_test.go`
-

## Tests

```
go test /dot/sync/ -short -v
go test /dot/sync/ -short -v --tags=integration
```

## Issues

- closes: #2023 

## Primary Reviewer

- @timwu20

### CODE REVIEW

**Comment by qdm12** [comment]
@edwardmack didn't git detect the file movements ('replace')? If you didn't modify the files much, git should had detected these are just file movements, drastically reducing deltas (correct me if I'm wrong though).
If you did modify the files enough that git doesn't detect them as file moves, maybe have a single commit moving all the files (even if compilation fails) and a second one modifying them, and we can then merge (not squash and merge) your PR so git stuff is kept in the development branch (who modified what etc.)

File: dot/sync/chain_sync_integeration_test.go [hunk 1]
```diff
@@ -31,8 +34,8 @@ const (
 )
 
 func newTestChainSync(t *testing.T) (*chainSync, *blockQueue) {
-	header, err := types.NewHeader(common.NewHash([]byte{0}),
-		trie.EmptyHash, trie.EmptyHash, big.NewInt(0), types.NewDigest())
+	header, err := types.NewHeader(common.NewHash([]byte{0}), trie.EmptyHash, trie.EmptyHash, big.NewInt(0),
+		types.NewDigest())
```
Comment (qdm12) [inline dot/sync/chain_sync_integeration_test.go:38]: Why these few changes? Splitting lines equally looked nice no? Just curious, don't bother re-changing though.

File: dot/sync/chain_sync_integeration_test.go [hunk 2]
```diff
@@ -109,11 +112,12 @@ func TestChainSync_SetPeerHead(t *testing.T) {
...
 	cs.blockState.(*syncmocks.BlockState).On("BestBlockHeader").Return(header, nil)
-	fin, err = types.NewHeader(common.NewHash([]byte{0}), trie.EmptyHash,
-		trie.EmptyHash, big.NewInt(1000), types.NewDigest())
+	fin, err = types.NewHeader(common.NewHash([]byte{0}), trie.EmptyHash, trie.EmptyHash, big.NewInt(1000),
+		types.NewDigest())
 	require.NoError(t, err)
 	cs.blockState.(*syncmocks.BlockState).On("GetHighestFinalisedHeader").Return(fin, nil)
-	cs.blockState.(*syncmocks.BlockState).On("GetHashByNumber", mock.AnythingOfType("*big.Int")).Return(common.Hash{}, nil)
+	cs.blockState.(*syncmocks.BlockState).On("GetHashByNumber", mock.AnythingOfType("*big.Int")).Return(common.
+		Hash{}, nil)
```
Comment (qdm12) [inline dot/sync/chain_sync_integeration_test.go:120]: Splitting `common.` and `Hash{}` ugh not prettyyy

File: dot/sync/chain_sync_integeration_test.go [hunk 3]
```diff
@@ -480,7 +484,7 @@ func TestValidateBlockData(t *testing.T) {
 	err = cs.validateBlockData(req, &types.BlockData{
 		Header: &types.Header{},
 	}, "")
-	require.ErrorIs(t, err, errNilBodyInResponse)
+	require.Equal(t, errNilBodyInResponse, err)
```
Comment (qdm12) [inline dot/sync/chain_sync_integeration_test.go:487]: Keep it as ErrorIs in case we wrap it in the future

**Review by qdm12 — APPROVED** [review]

File: dot/sync/disjoint_block_set_integeration_test.go [hunk 4]
```diff
@@ -204,24 +198,3 @@ func TestDisjointBlockSet_getReadyDescendants_blockNotComplete(t *testing.T) {
 	require.Equal(t, block1.ToBlockData(), ready[0])
 	require.Equal(t, block2.ToBlockData(), ready[1])
 }
-
-func TestDisjointBlockSet_ClearBlocks(t *testing.T) {
```
Comment (noot) [inline dot/sync/disjoint_block_set_integeration_test.go:208]: where did this test go?

File: dot/sync/chain_sync_integeration_test.go (hunk 2 above)
Comment (edwardmack) [reply dot/sync/chain_sync_integeration_test.go:120]: Right, my IDE did this, I've updated.

File: dot/sync/chain_sync_integeration_test.go (hunk 3 above)
Comment (edwardmack) [reply dot/sync/chain_sync_integeration_test.go:487]: Updated.

File: dot/sync/disjoint_block_set_integeration_test.go (hunk 4 above)
Comment (edwardmack) [reply dot/sync/disjoint_block_set_integeration_test.go:208]: right, I re-added this.

File: dot/sync/syncer_integeration_test.go [hunk 5]
```diff
@@ -163,69 +154,8 @@ func newTestGenesisWithTrieAndHeader(t *testing.T) (*genesis.Genesis, *trie.Trie
...
 
-	genesisHeader, err := types.NewHeader(common.NewHash([]byte{0}),
-		genTrie.MustHash(), trie.EmptyHash, big.NewInt(0), types.NewDigest())
+	genesisHeader, err := types.NewHeader(common.NewHash([]byte{0}), genTrie.MustHash(), trie.EmptyHash,
+		big.NewInt(0), types.NewDigest())
 	require.NoError(t, err)
 	return gen, genTrie, genesisHeader
 }
-
-func TestHighestBlock(t *testing.T) {
```
Comment (noot) [inline dot/sync/syncer_integeration_test.go:172]: can you re-add this?

File: dot/sync/tip_syncer_integeration_test.go [hunk 6]
```diff
@@ -176,10 +173,8 @@ func TestTipSyncer_handleTick_case1(t *testing.T) {
 			targetNumber: fin.Number,
 			direction:    network.Descending,
 			requestData:  bootstrapRequestData,
-			pendingBlock: s.pendingBlocks.getBlock(common.Hash{0xb}),
```
Comment (noot) [inline dot/sync/tip_syncer_integeration_test.go:179]: why is this removed? same with the others in this PR

File: dot/sync/syncer_integeration_test.go (hunk 5 above)
Comment (edwardmack) [reply dot/sync/syncer_integeration_test.go:172]: Yes, re-added.

File: dot/sync/tip_syncer_integeration_test.go (hunk 6 above)
Comment (edwardmack) [reply dot/sync/tip_syncer_integeration_test.go:179]: Honestly, I don't know why/how these got removed, probably during a rebase where I picked some wrong code to delete from conflict.  I've added these back in.

File: dot/sync/chain_sync_integeration_test.go [hunk 7]
```diff
@@ -332,7 +337,7 @@ func TestWorkerToRequests(t *testing.T) {
 					StartingBlock: *variadic.MustNewUint64OrHash(1),
 					EndBlockHash:  nil,
 					Direction:     network.Ascending,
-					Max:           &max128,
+					Max:           &max9,
```
Comment (noot) [inline dot/sync/chain_sync_integeration_test.go:340]: i think this was updated, can you pull development again?

File: dot/sync/chain_sync_integeration_test.go [hunk 8]
```diff
@@ -826,140 +826,3 @@ func TestChainSync_determineSyncPeers(t *testing.T) {
 	require.Equal(t, 1, len(peers))
 	require.Equal(t, []peer.ID{testPeerB}, peers)
 }
-
-func TestChainSync_highestBlock(t *testing.T) {
```
Comment (noot) [inline dot/sync/chain_sync_integeration_test.go:830]: this is still missing

File: dot/sync/chain_sync_integeration_test.go (hunk 7 above)
Comment (edwardmack) [reply dot/sync/chain_sync_integeration_test.go]: I've fixed this to match #2256.

File: dot/sync/chain_sync_integeration_test.go (hunk 8 above)
Comment (edwardmack) [reply dot/sync/chain_sync_integeration_test.go]: This has been re-added.

File: dot/sync/chain_sync_integeration_test.go [hunk 9]
```diff
@@ -41,8 +44,8 @@ func newTestChainSync(t *testing.T) (*chainSync, *blockQueue) {
 	bs.On("HasHeader", mock.AnythingOfType("common.Hash")).Return(true, nil)
 
 	net := new(syncmocks.Network)
-	net.On("DoBlockRequest", mock.AnythingOfType("peer.ID"),
-		mock.AnythingOfType("*network.BlockRequestMessage")).Return(nil, nil)
+	net.On("DoBlockRequest", mock.AnythingOfType("peer.ID"), mock.AnythingOfType("*network.BlockRequestMessage")).
+		Return(nil, nil)
```
Comment (qdm12) [inline dot/sync/chain_sync_integeration_test.go:48]: There are a lot of deltas due to line breaking up by your IDE, maybe could you disable this feature? This adds load for reviewers.
We already have the `lll` linter to ensure line don't go above 120chars.
And also reset back those changes for git blame purposes 😄

File: dot/sync/syncer_integeration_test.go [hunk 10]
```diff
@@ -58,29 +58,27 @@ func newMockBabeVerifier() *mocks.BabeVerifier {
...
 }
 
-//go:generate mockgen -destination=mock_telemetry_test.go -package $GOPACKAGE github.com/ChainSafe/gossamer/dot/telemetry Client
-
 func newTestSyncer(t *testing.T) *Service {
-	wasmer.DefaultTestLogLvl = 3
 	ctrl := gomock.NewController(t)
-	telemetryMock := NewMockClient(ctrl)
-	telemetryMock.EXPECT().SendMessage(gomock.Any()).AnyTimes()
+	defer ctrl.Finish()
```
Comment (qdm12) [inline dot/sync/syncer_integeration_test.go:68]: no need since go 1.16

```suggestion
```

**Comment by qdm12** [comment]
Also sorry to be the git maniac here, but perhaps you could interactively rebase on your first commit to have the following commits:

1. file movements / deletions
2. file changes

And then merge commits instead of squash and merge? That way file movements will be kept in history and easier to merge/rebase on for other branches (and also for git blame etc.). Otherwise git might just think it's purely new files and older files got deleted. I can help with that let me know :wink:

**Review by noot — APPROVED** [review]

File: dot/sync/syncer_integeration_test.go [hunk 11]
```diff
@@ -58,29 +58,27 @@ func newMockBabeVerifier() *mocks.BabeVerifier {
...
 
-	cfg := &Config{
-		Telemetry: telemetryMock,
-	}
+	mockTelemetryClient := NewMockClient(ctrl)
+	mockTelemetryClient.EXPECT().SendMessage(gomock.Any())
+
+	wasmer.DefaultTestLogLvl = 3
```
[the remark this answers is not in the document]
Comment (edwardmack) [reply dot/sync/syncer_integeration_test.go:72]: Updated.

File: dot/sync/syncer_integeration_test.go (hunk 10 above)
Comment (edwardmack) [reply dot/sync/syncer_integeration_test.go:67]: Updated.

File: dot/sync/tip_syncer_integeration_test.go [hunk 12]
```diff
@@ -214,7 +215,6 @@ func TestTipSyncer_handleTick_case2(t *testing.T) {
 	require.Equal(t, expected, w)
 	require.True(t, s.pendingBlocks.hasBlock(header.Hash()))
 }
-
```
[the remark this answers is not in the document]
Comment (edwardmack) [reply dot/sync/tip_syncer_integeration_test.go:216]: Not sure, I've re-added.

**Comment by edwardmack** [comment]
@qdm12 I've interactively rebased these commits, it seems to be clearer now.  Please take a look again and let me know your thoughts.

File: dot/sync/chain_sync_integeration_test.go [hunk 13]
```diff
@@ -40,8 +43,8 @@ func newTestChainSync(t *testing.T) (*chainSync, *blockQueue) {
...
```
Comment (qdm12) [inline dot/sync/chain_sync_integeration_test.go:47]: nit to keep it as close as it was for git file moves tracking

```suggestion
	net.On("DoBlockRequest", mock.AnythingOfType("peer.ID"), 
		mock.AnythingOfType("*network.BlockRequestMessage")).Return(nil, nil)
```

File: dot/sync/chain_sync_integeration_test.go [hunk 14]
```diff
@@ -657,8 +659,7 @@ func TestChainSync_doSync(t *testing.T) {
 	}
 
 	cs.network = new(syncmocks.Network)
-	cs.network.(*syncmocks.Network).On("DoBlockRequest",
-		mock.AnythingOfType("peer.ID"),
+	cs.network.(*syncmocks.Network).On("DoBlockRequest", mock.AnythingOfType("peer.ID"),
```
Comment (qdm12) [inline dot/sync/chain_sync_integeration_test.go:662]: nit to keep it as close as it was for git file moves tracking

```suggestion
	cs.network.(*syncmocks.Network).On("DoBlockRequest",
		mock.AnythingOfType("peer.ID"),
```

File: dot/sync/chain_sync_integeration_test.go [hunk 15]
```diff
@@ -693,8 +692,7 @@ func TestChainSync_doSync(t *testing.T) {
 	// test to see if descending blocks get reversed
 	req.Direction = network.Descending
 	cs.network = new(syncmocks.Network)
-	cs.network.(*syncmocks.Network).On("DoBlockRequest",
-		mock.AnythingOfType("peer.ID"),
+	cs.network.(*syncmocks.Network).On("DoBlockRequest", mock.AnythingOfType("peer.ID"),
```
Comment (qdm12) [inline dot/sync/chain_sync_integeration_test.go:695]:
```suggestion
	cs.network.(*syncmocks.Network).On("DoBlockRequest",
		mock.AnythingOfType("peer.ID"),
```

File: dot/sync/disjoint_block_set_integeration_test.go [hunk 16]
```diff
@@ -15,9 +18,6 @@ import (
 
 func TestDisjointBlockSet(t *testing.T) {
 	s := newDisjointBlockSet(pendingBlocksLimit)
-	s.timeNow = func() time.Time {
-		return time.Time{}
-	}
```
Comment (qdm12) [inline dot/sync/disjoint_block_set_integeration_test.go:20]: Are you sure this is no longer needed?

File: dot/sync/message_integeration_test.go [hunk 17]
```diff
@@ -32,7 +35,7 @@ func addTestBlocksToState(t *testing.T, depth uint, blockState BlockState) {
 				ParentHash: previousHash,
 				Number:     previousNum + i,
 				StateRoot:  trie.EmptyHash,
-				Digest:     digest,
+				Digest:     types.NewDigest(),
```
Comment (qdm12) [inline dot/sync/message_integeration_test.go:38]: Why was this changed from `digest` (with that `prd` added)?

If there is reason, maybe remove the `digest` variable since it's not used I think.

File: dot/sync/syncer_integeration_test.go [hunk 18]
```diff
@@ -57,29 +57,26 @@ func newMockBabeVerifier() *mocks.BabeVerifier {
 
 func newMockNetwork() *mocks.Network {
 	m := new(mocks.Network)
-	m.On("DoBlockRequest", mock.AnythingOfType("peer.ID"),
-		mock.AnythingOfType("*network.BlockRequestMessage")).Return(nil, nil)
+	m.On("DoBlockRequest", mock.AnythingOfType("peer.ID"), mock.AnythingOfType("*network.BlockRequestMessage")).
+		Return(nil, nil)
```
Comment (qdm12) [inline dot/sync/syncer_integeration_test.go:61]:
```suggestion
	m.On("DoBlockRequest", mock.AnythingOfType("peer.ID"), 
		mock.AnythingOfType("*network.BlockRequestMessage")).Return(nil, nil)
```

File: dot/sync/syncer_integeration_test.go [hunk 19]
```diff
@@ -124,24 +121,23 @@ func newTestSyncer(t *testing.T) *Service {
...
-			err = stateSrvc.Block.AddBlock(block)
-			require.NoError(t, err)
-
-			stateSrvc.Block.StoreRuntime(block.Header.Hash(), instance)
-			logger.Debugf("imported block %s and stored state trie with root %s",
-				block.Header.Hash(), ts.MustRoot())
-			return nil
-		})
+	cfg.BlockImportHandler.(*mocks.BlockImportHandler).On("HandleBlockImport", mock.AnythingOfType("*types.Block"),
+		mock.AnythingOfType("*storage.TrieState")).Return(func(block *types.Block, ts *rtstorage.TrieState) error {
```
Comment (qdm12) [inline dot/sync/syncer_integeration_test.go:125]: keeping it as it was

```suggestion
	cfg.BlockImportHandler.(*mocks.BlockImportHandler).On(
		"HandleBlockImport", mock.AnythingOfType("*types.Block"), mock.AnythingOfType("*storage.TrieState")).
		Return(func(block *types.Block, ts *rtstorage.TrieState) error {
```

**Review by qdm12 — COMMENTED** [review]
Nice rebase! Just a few questions and suggestions to revert back to how it was to reduce deltas.

We should be able to squash and merge and git will track file movements correctly, so don't worry about adding commits :wink:

File: dot/sync/chain_sync_integeration_test.go (hunk 13 above)
Comment (edwardmack) [reply dot/sync/chain_sync_integeration_test.go:47]: Updated.

File: dot/sync/chain_sync_integeration_test.go (hunk 14 above)
Comment (edwardmack) [reply dot/sync/chain_sync_integeration_test.go:662]: Updated.

File: dot/sync/chain_sync_integeration_test.go (hunk 15 above)
Comment (edwardmack) [reply dot/sync/chain_sync_integeration_test.go:695]: Updated.

File: dot/sync/message_integeration_test.go (hunk 17 above)
Comment (edwardmack) [reply dot/sync/message_integeration_test.go:38]: Not sure why this was changed, changed it back and confirmed test that use this are working.

File: dot/sync/syncer_integeration_test.go (hunk 18 above)
Comment (edwardmack) [reply dot/sync/syncer_integeration_test.go:61]: Updated.

File: dot/sync/syncer_integeration_test.go (hunk 19 above)
Comment (edwardmack) [reply dot/sync/syncer_integeration_test.go:125]: Update.

File: dot/sync/disjoint_block_set_integeration_test.go (hunk 16 above)
Comment (edwardmack) [reply dot/sync/disjoint_block_set_integeration_test.go]: Re-added, and fixed test.

File: dot/sync/chain_sync_integeration_test.go [hunk 20]
```diff
@@ -647,7 +650,6 @@ func TestChainSync_doSync(t *testing.T) {
 	resp := &network.BlockResponseMessage{
 		BlockData: []*types.BlockData{
 			{
-				Hash: common.Hash{0x1},
```
Comment (qdm12) [inline dot/sync/chain_sync_integeration_test.go:650]: Are we sure about these removals? Does the test pass locally?

**Review by qdm12 — APPROVED** [review]

**Review by qdm12 — CHANGES_REQUESTED** [review]
Can you squash the last few commits in the 2nd commit, and then let's **rebase and merge**

**Review by qdm12 — APPROVED** [review]
Ignore the comment above; changes are minimal so the squash and merge should track file movements properly. Let's just squash and merge it :100:

**Review by qdm12 — CHANGES_REQUESTED** [review]
Please add `dot/sync` to https://github.com/ChainSafe/gossamer/blob/19a2b4e36d8f46b062aa1b139edc514f12f119f5/.github/workflows/integration-tests.yml#L28 and then let's merge this! :+1:

**Comment by edwardmack** [comment]
Added `dot/sync` package to github integration workflow.

File: dot/sync/disjoint_block_set_integeration_test.go [hunk 21]
```diff
@@ -40,7 +43,6 @@ func TestDisjointBlockSet(t *testing.T) {
 	s.addHeader(header)
 	require.True(t, s.hasBlock(header.Hash()))
 	require.Equal(t, 2, s.size())
-
```
Comment (kishansagathiya) [inline dot/sync/disjoint_block_set_integeration_test.go:43]: ?

File: dot/sync/tip_syncer_integeration_test.go [hunk 22]
```diff
@@ -178,7 +180,6 @@ func TestTipSyncer_handleTick_case1(t *testing.T) {
 			pendingBlock: s.pendingBlocks.getBlock(common.Hash{0xb}),
 		},
 	}
-
```
Comment (kishansagathiya) [inline dot/sync/tip_syncer_integeration_test.go:181]: ?

File: dot/sync/chain_processor_integration_test.go [hunk 23]
```diff
@@ -206,7 +209,7 @@ func TestChainProcessor_HandleJustification(t *testing.T) {
 	d, err := types.NewBabeSecondaryPlainPreDigest(0, 1).ToPreRuntimeDigest()
 	require.NoError(t, err)
 	digest := types.NewDigest()
-	err = digest.Add(*d)
```
Comment (kishansagathiya) [inline dot/sync/chain_processor_integration_test.go:209]: Why has this changed?

`digest.Add` takes an interface, but I think if you look at its code, you will find that it accepts a value.

**Review by kishansagathiya — CHANGES_REQUESTED** [review]
Just a few comments

- Make sure to not modify irrelevant code.
- Move back the line spacing that was removed.

File: dot/sync/disjoint_block_set_integeration_test.go (hunk 21 above)
Comment (edwardmack) [reply dot/sync/disjoint_block_set_integeration_test.go:43]: Re-added line spacing.

File: dot/sync/tip_syncer_integeration_test.go (hunk 22 above)
Comment (edwardmack) [reply dot/sync/tip_syncer_integeration_test.go:181]: re-added line spacing.

**Comment by edwardmack** [comment]
@kishansagathiya I've addressed PR comments, please re-review and let me know if any issues remain.

**Review by kishansagathiya — APPROVED** [review]
:tada:

File: .github/workflows/integration-tests.yml [hunk 24]
```diff
@@ -25,6 +25,7 @@ jobs:
           [
             github.com/ChainSafe/gossamer/dot/rpc/modules,
             github.com/ChainSafe/gossamer/lib/babe,
+            github.com/ChainSafe/gossamer/dot/sync,
```
Comment (qdm12) [inline .github/workflows/integration-tests.yml:28]: You might want to rebase your branch on development to get the missing required workflows for the `dot` and `dot/core` package integration tests

**Review by qdm12 — CHANGES_REQUESTED** [review]

File: .github/workflows/integration-tests.yml (hunk 24 above)
Comment (edwardmack) [reply .github/workflows/integration-tests.yml:30]: Good point, rebased.

File: dot/sync/syncer_integeration_test.go [hunk 25]
```diff
@@ -62,24 +62,21 @@ func newMockNetwork() *mocks.Network {
 	return m
 }
 
-//go:generate mockgen -destination=mock_telemetry_test.go -package $GOPACKAGE github.com/ChainSafe/gossamer/dot/telemetry Client
-
```
Comment (qdm12) [inline dot/sync/syncer_integeration_test.go:66]: Why is this removed? Don't we need it? I can't find another one in the `dot/sync` package.

**Review by qdm12 — CHANGES_REQUESTED** [review]

File: dot/sync/syncer_integeration_test.go (hunk 25 above)
Comment (edwardmack) [reply dot/sync/syncer_integeration_test.go:66]: This shouldn't have been removed, I've re-added.

File: dot/sync/chain_sync_integeration_test.go (hunk 20 above)
Comment (edwardmack) [reply dot/sync/chain_sync_integeration_test.go:650]: I've re-tested, and I have re-added these.  I'm not sure why I removed these.

**Review by timwu20 — APPROVED** [review]

**Review by qdm12 — APPROVED** [review]
